### PR TITLE
Add user directory tests

### PR DESCRIPTION
**Pull request by erikjohnston**

### CODE REVIEW

File: tests/10apidoc/01register.pl [hunk 1]
```diff
@@ -82,11 +82,16 @@
 
 my $next_anon_uid = 1;
 
+sub create_localpart
```
Comment (leonerd) [inline tests/10apidoc/01register.pl]: Tiny nit: I'd be tempted to call this `sprintf_localpart` or somesuch, to point out it just does some string formatting and doesn't e.g. talk to a Matrix server and Do Things.

**Comment by erikjohnston** [comment]
(I added a test for changing profiles too, sorry)

File: tests/52user-directory.pl [hunk 2]
```diff
@@ -0,0 +1,449 @@
...
+         matrix_create_room( $user,
+            preset => "public_chat",
+         )
+      })->then( sub {
+         ( $room_id ) = @_;
+
+         try_repeat_until_success( sub {
+            do_request_json_for( $user,
+               method  => "POST",
+               uri     => "/r0/user_directory/search",
```
Comment (leonerd) [inline tests/52user-directory.pl]: Is this really in `/r0` already?

File: tests/52user-directory.pl [hunk 3]
```diff
@@ -0,0 +1,449 @@
...
+               any { $_->{user_id} eq $user->user_id } @{ $body->{results} }
+                  or die "user not in list";
+
+               Future->done( 1 )
+            })
+         })
+      });
+   };
+
+test "User in private room doesn't appears in user directory",
```
Comment (leonerd) [inline tests/52user-directory.pl]: Smallest of grammatical nits - "appear"

File: tests/52user-directory.pl [hunk 4]
```diff
@@ -0,0 +1,449 @@
...
+         any { $_->{user_id} eq $user->user_id } @{ $body->{results} }
+            and die "user in list";
+
+         pass "User not in directory after history_visibility set to shared";
+
+         Future->done( 1 )
+      });
+   };
+
+test "User in remote room don't appear in user directory after server left room",
```
Comment (leonerd) [inline tests/52user-directory.pl]: "doesn't"?

File: tests/52user-directory.pl [hunk 5]
```diff
@@ -0,0 +1,449 @@
...
+         log_if_fail "Second Body", $body;
+
+         any { $_->{user_id} eq $user->user_id } @{ $body->{results} }
+            or die "user not in list";
+
+         Future->done( 1 )
+      });
+   };
+
+sub create_and_set_random_displayname
```
Comment (leonerd) [inline tests/52user-directory.pl]: Despite its convenience, this all-in-one function does feel a bit awkward to use, for having to capture the `$displayname` each time. I wonder about splitting it in two, so you can

    my $displayname = generate_random_displayname();
    ...
    matrix_set_displayname( $user, $displayname )->then ...

**Comment by leonerd** [comment]
Overall not bad but it's missing quite a lot of trailing `;`s on the final statements in blocks. Perl doesn't strictly complain about their lack but it's usual style to include them in regular code flow.

**Comment by erikjohnston** [comment]
@leonerd PTAL :)

File: tests/52user-directory.pl [hunk 6]
```diff
@@ -0,0 +1,437 @@
...
+                  content => { join_rule => "public" },
+               );
+            } else {
+               matrix_put_room_state( $creator, $room_id,
+                  type    => "m.room.history_visibility",
+                  content => { history_visibility => "world_readable" },
+               );
+            }
+         })->then( sub {
+            get_user_dir_synced( $user, $displayname )
```
Comment (leonerd) [inline tests/52user-directory.pl]: Still missing `;`

File: tests/52user-directory.pl [hunk 7]
```diff
@@ -0,0 +1,437 @@
...
+                  content => { join_rule => "invite" },
+               );
+            } else {
+               matrix_put_room_state( $creator, $room_id,
+                  type    => "m.room.history_visibility",
+                  content => { history_visibility => "shared" },
+               );
+            }
+         })->then( sub {
+            get_user_dir_synced( $user, $displayname )
```
Comment (leonerd) [inline tests/52user-directory.pl]: `;`

File: tests/52user-directory.pl [hunk 8]
```diff
@@ -0,0 +1,437 @@
...
+            type    => "m.room.join_rules",
+            content => { join_rule => "public" },
+         );
+      })->then( sub {
+         matrix_put_room_state( $creator, $room_id,
+            type    => "m.room.history_visibility",
+            content => { history_visibility => "world_readable" },
+         );
+      })->then( sub {
+         get_user_dir_synced( $user, $displayname )
```
Comment (leonerd) [inline tests/52user-directory.pl]: `;`

File: tests/52user-directory.pl [hunk 9]
```diff
@@ -0,0 +1,437 @@
...
+         get_user_dir_synced( $user, $displayname );
+      })->then( sub {
+         my ( $body ) = @_;
+
+         log_if_fail "Body", $body;
+
+         any { $_->{user_id} eq $user->user_id } @{ $body->{results} }
+            or die "user not in list";
+
+         matrix_set_displayname( $user, $second_displayname )
```
Comment (leonerd) [inline tests/52user-directory.pl]: `;`

**Comment by leonerd** [comment]
Looking much better now. Still a few missing `;`s but other than that LGTM.

**Pull request merged**